File: sdk/desktop/windows/tests/Kseal.Desktop.Tests/FakeHardwareKeyStore.cs
namespace Kseal.Desktop.Tests;

/// <summary>
/// Deterministic test fake for <see cref="IHardwareKeyStore"/>: prepends a magic
/// header and XOR-obfuscates the payload so a "sealed" blob is distinguishable
/// from the plaintext (a foreign/legacy blob is rejected on unseal). Lets the
/// proof-key sealing logic be exercised on any host without a TPM.
/// </summary>
internal sealed class FakeHardwareKeyStore : IHardwareKeyStore
{
    private static readonly byte[] Magic = "KSEA"u8.ToArray();
    private const byte XorMask = 0x5A;

    public bool IsHardwareBacked => true;

    public byte[] Seal(byte[] plaintext)
    {
        byte[] outBuf = new byte[Magic.Length + plaintext.Length];
        Magic.CopyTo(outBuf, 0);
        for (int i = 0; i < plaintext.Length; i++) outBuf[Magic.Length + i] = (byte)(plaintext[i] ^ XorMask);
        return outBuf;
    }

    public byte[] Unseal(byte[] sealedBlob)
    {
        if (sealedBlob.Length < Magic.Length || !sealedBlob.AsSpan(0, Magic.Length).SequenceEqual(Magic))
        {
            throw new HardwareKeyStoreException("not a blob produced by this store");
        }
        byte[] outBuf = new byte[sealedBlob.Length - Magic.Length];
        for (int i = 0; i < outBuf.Length; i++) outBuf[i] = (byte)(sealedBlob[Magic.Length + i] ^ XorMask);
        return outBuf;
    }
}

/// <summary>A store whose operations always fail, to test graceful software degradation.</summary>
internal sealed class FailingHardwareKeyStore : IHardwareKeyStore
{
    public bool IsHardwareBacked => true;
    public byte[] Seal(byte[] plaintext) => throw new HardwareKeyStoreException("seal unavailable");
    public byte[] Unseal(byte[] sealedBlob) => throw new HardwareKeyStoreException("unseal unavailable");
}

File: sdk/desktop/macos/Tests/KsealDesktopTests/FakeHardwareKeyStore.swift
import Foundation
@testable import KsealDesktop

/// Deterministic stand-in for a hardware-backed key store, used to exercise the
/// proof-key sealing/migration logic on any host. `seal` prepends a magic header
/// and XOR-obfuscates the payload; `unseal` validates the header so a foreign /
/// legacy-raw blob fails to unseal exactly like a real wrong-device ciphertext.
final class FakeHardwareKeyStore: HardwareKeyStore {
    static let magic: [UInt8] = [0x4B, 0x53, 0x45, 0x41] // "KSEA"
    private let mask: UInt8

    let isHardwareBacked: Bool

    init(isHardwareBacked: Bool = true, mask: UInt8 = 0x5A) {
        self.isHardwareBacked = isHardwareBacked
        self.mask = mask
    }

    func seal(_ plaintext: Data) throws -> Data {
        var out = Data(Self.magic)
        out.append(contentsOf: plaintext.map { $0 ^ mask })
        return out
    }

    func unseal(_ sealed: Data) throws -> Data {
        guard sealed.count >= Self.magic.count,
              Array(sealed.prefix(Self.magic.count)) == Self.magic else {
            throw HardwareKeyStoreError(message: "not sealed by this store")
        }
        return Data(sealed.dropFirst(Self.magic.count).map { $0 ^ mask })
    }
}

/// A store whose seal always fails, to drive the graceful software-degradation
/// path in the proof-key provider.
final class FailingHardwareKeyStore: HardwareKeyStore {
    let isHardwareBacked = true
    func seal(_ plaintext: Data) throws -> Data {
        throw HardwareKeyStoreError(message: "seal unavailable")
    }
    func unseal(_ sealed: Data) throws -> Data {
        throw HardwareKeyStoreError(message: "unseal unavailable")
    }
}
